YogaKit/Source/YGLayout.m

```diff
@@ -401,44 +406,86 @@ static CGFloat YGRoundPixelValue(CGFloat value)
   return roundf(value * scale) / scale;
 }
 
-static void YGApplyLayoutToViewHierarchy(UIView *view, BOOL preserveOrigin)
+static void YGApplyLayoutToHierarchy(id<YGLayoutEntity> entity, BOOL preserveOrigin, CGPoint offset)
 {
   NSCAssert([NSThread isMainThread], @"Framesetting should only be done on the main thread.");
 
-  const YGLayout *yoga = view.yoga;
+  const YGLayout *yoga = entity.yoga;
 
-  if (!yoga.isIncludedInLayout) {
+  if (!yoga || !yoga.isIncludedInLayout) {
      return;
   }
 
-  YGNodeRef node = yoga.node;
   const CGPoint topLeft = {
-    YGNodeLayoutGetLeft(node),
-    YGNodeLayoutGetTop(node),
-  };
-
-  const CGPoint bottomRight = {
-    topLeft.x + YGNodeLayoutGetWidth(node),
-    topLeft.y + YGNodeLayoutGetHeight(node),
+    YGNodeLayoutGetLeft(yoga->_node) + offset.x,
+    YGNodeLayoutGetTop(yoga->_node) + offset.y,
   };
 
-  const CGPoint origin = preserveOrigin ? view.frame.origin : CGPointZero;
-  view.frame = (CGRect) {
+  const CGPoint origin = preserveOrigin ? entity.frame.origin : CGPointZero;
+  entity.frame = (CGRect) {
     .origin = {
       .x = YGRoundPixelValue(topLeft.x + origin.x),
       .y = YGRoundPixelValue(topLeft.y + origin.y),
     },
     .size = {
-      .width = YGRoundPixelValue(bottomRight.x) - YGRoundPixelValue(topLeft.x),
-      .height = YGRoundPixelValue(bottomRight.y) - YGRoundPixelValue(topLeft.y),
+      .width = YGNodeLayoutGetWidth(yoga->_node),
+      .height = YGNodeLayoutGetHeight(yoga->_node),
     },
   };
 
   if (!yoga.isLeaf) {
-    for (NSUInteger i=0; i<view.subviews.count; i++) {
-      YGApplyLayoutToViewHierarchy(view.subviews[i], NO);
+    for (NSUInteger i=0; i<entity.subEntities.count; i++) {
+      YGApplyLayoutToHierarchy(entity.subEntities[i],
+                               NO,
+                               [entity conformsToProtocol:@protocol(UICoordinateSpace)] ? CGPointZero : entity.frame.origin);
```

Comment:
Ouch, checking protocol conformance is a pretty expensive operation to constantly be doing thru a node hierarchy on main thread. Any way we can optimize this?